### PR TITLE
Add TLS module

### PR DESCRIPTION
**Pull request by repeatedly**
**Which issue(s) this PR fixes**: 
- Fix bug of https://groups.google.com/forum/#!topic/fluentd/B818cI-OHdU

**What this PR does / why we need it**: 
Add TLS module to unify fluentd code.
In addition, support recent SSLContext changes about TLS version

TODO: Add test soon

**Docs Changes**:
Add `min_version`/`max_version` to TLS transport.

**Release Note**: 
- Add TLS module to unify code
- Allow TLS v1.3 when available
- Fix bug of TLS version setting

### CODE REVIEW

**Comment by repeatedly** [comment]
I searched the 3rd party plugins and `Fluent::PluginHelper::Socket::TLS_SUPPORTED_VERSION`/other constants are not used. We can remove it.

File: lib/fluent/tls.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+module Fluent
+  module TLS
+    DEFAULT_VERSION = :'TLSv1_2'
+    SUPPORTED_VERSIONS = [:'TLSv1_1', :'TLSv1_2', :'TLS1_1', :'TLS1_2']
+    ### follow httpclient configuration by nahi
+    # OpenSSL 0.9.8 default: "ALL:!ADH:!LOW:!EXP:!MD5:+SSLv2:@STRENGTH"
+    CIPHERS_DEFAULT = "ALL:!aNULL:!eNULL:!SSLv2" # OpenSSL >1.0.0 default
+
+    METHODS_MAP = {
+      TLSv1: OpenSSL::SSL::TLS1_VERSION,
```
Comment (repeatedly) [inline lib/fluent/tls.rb]: ruby 2.4 doesn't have these constants. Fix soon.

File: lib/fluent/tls.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        case
+        when min_version.nil? && max_version.nil?
+          min_version = METHODS_MAP[version]
+          max_version = METHODS_MAP[version]
+        when min_version.nil? && max_version
+          raise Fluentd::ConfigError, "When you set max_version, must set min_version together"
+        when min_version && max_version.nil?
+          raise Fluentd::ConfigError, "When you set min_version, must set max_version together"
+        else
+          min_version = METHODS_MAP[min_version] || min_version
```
Comment (ganmacs) [inline lib/fluent/tls.rb]: if the `METHODS_MAP[min_version]` returns `nil`,  a user might make a typo or another mistake. 
How about warning it?

Comment (repeatedly) [reply lib/fluent/tls.rb]: SSLContext raising an error for such cases.

**Comment by repeatedly** [comment]
Test passed

**Comment by repeatedly** [comment]
@ganmacs Check this again ;)

File: test/test_tls.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    'New TLS v1.1' => :'TLS1_1',
+    'Old TLS v1.1' => :'TLSv1_1',
+  }
+  TEST_TLS1_2_CASES = {
+    'New TLS v1.2' => :'TLS1_2',
+    'Old TLS v1.2' => :'TLSv1_2'
+  } 
+  TEST_TLS_CASES = TEST_TLS1_1_CASES.merge(TEST_TLS1_2_CASES)
+
+  sub_test_case 'constants' do
```
Comment (ganmacs) [inline test/test_tls.rb]: Do these tests need? It looks that they do nothing...

File: test/test_tls.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      }
+    end
+
+    data(TEST_TLS_CASES)
+    test 'can specify old/new syntax to min_version/max_version' do |ver|
+      omit "min_version=/max_version= is not supported" unless Fluent::TLS::MIN_MAX_AVAILABLE
+
+      assert_nothing_raised {
+        Fluent::TLS.set_version_to_context(@ctx, Fluent::TLS::DEFAULT_VERSION, ver, ver)
+      }
```
Comment (ganmacs) [inline test/test_tls.rb]: I think it needs to assert `@ctx.min_version` and `@ctx.max_version` is set the given value(Does what the method does set min and max version of TSL?).
The test is just checking this test does not raise an error and it doesn't check the behaivor of Fluent::TLS.set_version_to_context.

File: lib/fluent/tls.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      else
+        ctx.ssl_version = METHODS_MAP[version] || version
+      end
+
+      ctx
+    end
+    module_function :set_version_to_context
+  end
+end
+
```
Comment (ganmacs) [inline lib/fluent/tls.rb]: [nits] unnecessary line.

File: test/test_tls.rb (hunk 4 above)
Comment (repeatedly) [reply test/test_tls.rb]: > I think it needs to assert @ctx.min_version and @ctx.max_version is set the given value(Does what the method does set min and max version of TSL?).

I agree. But the problem is `SSLContext` doesn't provide accessor for `min_version`/`max_version`. If openssl gem implements it, we can add it.

File: test/test_tls.rb (hunk 3 above)
Comment (repeatedly) [reply test/test_tls.rb]: These constants are used by plugins.
These tests are for preventing unexpected value changes.

File: lib/fluent/tls.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                      TLSv1_1: OpenSSL::SSL::TLS1_1_VERSION,
+                      TLSv1_2: OpenSSL::SSL::TLS1_2_VERSION
+                    }
+                    map[:'TLSv1_3'] = OpenSSL::SSL::TLS1_3_VERSION if defined?(OpenSSL::SSL::TLS1_3_VERSION)
+                    MIN_MAX_AVAILABLE = true
+                    map.freeze
+                  rescue NameError
+                    # ruby 2.4 doesn't have OpenSSL::SSL::TLSXXX constants and min_version=/max_version= methods
+                    map = {
+                      TLS1: :'TLSv1',
```
Comment (cosmo0920) [inline lib/fluent/tls.rb]: There is map keys glitch between min max available case and unavailable cases.
Is it intended?
If this glitch is intended, we might have to write documentation about it.

Comment (repeatedly) [reply lib/fluent/tls.rb]: > we might have to write documentation about it.

This `METHODS_MAP` is private constant, so no public usecase.
Plugin developers don't need the details.

Comment (cosmo0920) [reply lib/fluent/tls.rb]: Oh! I see.

**Comment by repeatedly** [comment]
Add comment for current SSLContext test

**Comment by repeatedly** [comment]
If no other comment, I will merge the patch and release fluentd v1.9.1.

**Pull request merged**